services/wfm/src/main/java/org/openkilda/wfm/topology/nbworker/services/ValidationService.java

```diff
@@ -0,0 +1,120 @@
+ *   limitations under the License.
+ */
+
+package org.openkilda.wfm.topology.nbworker.services;
+
+import org.openkilda.messaging.info.rule.FlowEntry;
+import org.openkilda.messaging.info.rule.SwitchFlowEntries;
+import org.openkilda.messaging.info.switches.SyncRulesResponse;
+import org.openkilda.messaging.model.SwitchId;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import org.neo4j.driver.v1.Session;
+import org.neo4j.driver.v1.StatementResult;
+import org.neo4j.driver.v1.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class ValidationService {
+    private static final Logger logger = LoggerFactory.getLogger(ValidationService.class);
+    public static final Set<Long> SERVICE_RULES =
+            ImmutableSet.of(0x8000000000000001L, 0x8000000000000002L, 0x8000000000000003L);
+    private SwitchFlowEntries request;
+    private Session session;
+    private SwitchId switchId;
+
+    public ValidationService(SwitchFlowEntries request, Session session) {
+        this.request = request;
+        this.session = session;
+        switchId = request.getSwitchId();
+    }
+
+    /**
+     * Runs validating process.
+     * @return validation result as RulesResponse
+     */
+    public SyncRulesResponse validate() {
+        logger.debug("Validating rules on switch {}", switchId);
+
+        Set<Long> expectedCookies = readSegmentsCookies();
+        expectedCookies.addAll(readIngressCookies());
+        logger.debug("Expected rules on switch {}: {}", switchId, expectedCookies);
+
+        Set<Long> presentedCookies = request.getFlowEntries().stream()
+                .map(FlowEntry::getCookie)
+                .collect(Collectors.toSet());
+        presentedCookies.removeAll(SERVICE_RULES);
+        logger.debug("Presented rules on switch {}: {}", switchId, presentedCookies);
+
+        return makeRulesResponse(expectedCookies, presentedCookies, switchId);
+    }
+
+    private Set<Long> readSegmentsCookies() {
+        String q = "MATCH (:switch)-[fs:flow_segment]->(sw:switch) WHERE sw.name=$switch_id RETURN fs";
+        StatementResult queryResults = session.run(q, ImmutableMap.of("switch_id", switchId.toString()));
+        return extractCookies(queryResults, "fs");
+    }
+
+    private Set<Long> readIngressCookies() {
+        String q = "MATCH (sw:switch)-[r:flow]->(:switch) WHERE sw.name=$switch_id RETURN r";
+        StatementResult queryResults = session.run(q, ImmutableMap.of("switch_id", switchId.toString()));
+        return extractCookies(queryResults, "r");
+    }
+
+    private Set<Long> extractCookies(StatementResult queryResults, String fieldName) {
+        return queryResults.list()
+                .stream()
+                .map(record -> Optional.ofNullable(record)
```

Comment:
Ups... looks like I have pointed you in wrong direction. `.get()` method will not return "real" null, it will return some wrapper and this wrapper will throw one of `org.neo4j.driver.v1.exceptions.value.ValueException` for subsequent `.get` or `.asLong()` calls.

Reply:
Ohh. I need to test it manually.